src/dataFiles.ts

```diff
@@ -1,8 +1,161 @@
+import { memDB } from "./db.js";
 import { logger } from "./logger.js";
 import { getRuntimeConfig } from "./runtimeConfig.js";
+import {
+	createSearcheeFromPath,
+	File,
+	getFilesFromDataRoot,
+	parseTitle,
+	SearcheeWithoutInfoHash,
+} from "./searchee.js";
+import { createEnsemblePieces, EnsembleEntry } from "./torrent.js";
+import { createKeyTitle } from "./utils.js";
+import { isOk } from "./Result.js";
+
+interface DataEntry {
+	title: string;
+	path: string;
+}
+
+const watchers: Map<string, FSWatcher | null> = new Map();
+
+function startWatcher(dataDir: string): FSWatcher {
+	return watch(dataDir, { recursive: true, persistent: false }, () => {
+		watchers.get(dataDir)!.close();
+		watchers.set(dataDir, null);
+	});
+}
+
+export async function initializeDataDirs(): Promise<void> {
+	const { dataDirs } = getRuntimeConfig();
+	if (await memDB.schema.hasTable("data")) return;
+	await memDB.schema.createTable("data", (table) => {
+		table.string("path").primary();
+		table.string("title");
+	});
+	if (!dataDirs?.length) return;
+	for (const dataDir of dataDirs) {
+		watchers.set(dataDir, null);
+	}
+	logger.info("Indexing dataDirs for reverse lookup...");
+	await indexDataDirs();
+}
+
+async function indexDataDir(
+	dataDir: string,
+	prevWatcher: FSWatcher | null,
+): Promise<void> {
+	const { maxDataDepth, seasonFromEpisodes } = getRuntimeConfig();
+	if (prevWatcher) return;
+	logger.verbose(`Indexing ${dataDir} due to recent changes...`);
+	watchers.set(dataDir, startWatcher(dataDir));
+	const memoizedPaths = new Map<string, string[]>();
+	const memoizedLengths = new Map<string, number>();
+	for (const dirent of readdirSync(dataDir)) {
+		const entry = join(dataDir, dirent);
+		const paths = findPotentialNestedRoots(entry, maxDataDepth);
+		const dataRows: DataEntry[] = [];
+		const ensembleRows: EnsembleEntry[] = [];
+		for (const path of paths) {
+			const files = getFilesFromDataRoot(
+				path,
+				memoizedPaths,
+				memoizedLengths,
+			);
+			const title = parseTitle(basename(path), files, path);
+			if (!title) continue;
+			dataRows.push({ title, path });
+			if (seasonFromEpisodes) {
+				const ensembleEntry = await indexEnsembleDataEntry(
+					title,
+					path,
+					files,
+				);
+				if (ensembleEntry) ensembleRows.push(ensembleEntry);
+			}
+		}
+		const batchSize = 100;
+		for (let i = 0; i < dataRows.length; i += batchSize) {
+			const batch = dataRows.slice(i, i + batchSize);
+			if (!batch.length) break;
+			await memDB("data").insert(batch).onConflict("path").ignore();
+		}
+		for (let i = 0; i < ensembleRows.length; i += batchSize) {
+			const batch = ensembleRows.slice(i, i + batchSize);
+			if (!batch.length) break;
+			await memDB("ensemble").insert(batch).onConflict("path").ignore();
+		}
+	}
+}
+
+export async function indexDataDirs(): Promise<void> {
+	await Promise.all(
+		Array.from(watchers.entries()).map(([dataDir, prevWatcher]) =>
+			indexDataDir(dataDir, prevWatcher),
```

Comment:
That was the watcher for each dataDir. It can either be null, meaning it found something and we should reindex, or FSWatcher which means we can skip indexing.